client/network/test/src/sync.rs

```diff
@@ -616,9 +616,10 @@ fn syncs_header_only_forks() {
 	let small_hash = net.peer(0).client().info().best_hash;
 	net.peer(1).push_blocks(4, false);
 
-	net.block_until_sync();
 	// Peer 1 will sync the small fork even though common block state is missing
-	assert!(net.peer(1).has_block(&small_hash));
+	while !net.peer(1).has_block(&small_hash) {
+		net.block_until_idle();
```

Comment:
It is probably worth to consider using a Future with a timeout for all those `block_until_*()` methods.

Reply:
Why?

Reply:
A test that never ends, still shows that something is broken.

Reply:
And how do we tell the difference between a really slow test and a test that never ends :thinking: 

Although I would assume that the CI has some sort of timeouts in place. So a test run would be killed at some point in time and then we are back to the above question.

Reply:
For sure CI has a timeout...

> And how do we tell the difference between a really slow test and a test that never ends

The difference is that the later never ends... 

> So a test run would be killed at some point in time and then we are back to the above question.

We are not back to this question. Does this test looks like it runs multiple hours? No! Exactly and that solves the problem.

We don't need to discuss this here. Timeouts are a very bad idea for tests, because they introduce flakiness. I have seen enough of these tests that failed because of timeouts that wasn't long enough because the load was too high.